### PR TITLE
Include PartitionMappings on ExternalAssetNode

### PR DESCRIPTION
**Pull request by sryza**
### Summary & Motivation

This will enable the asset reconciliation sensor to become the asset reconciliation daemon and allow projected logical version logic in host processes to work with partitions.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 from dagster._core.definitions.partition_key_range import PartitionKeyRange
 from dagster._core.definitions.partition_mapping import PartitionMapping
 from dagster._core.definitions.time_window_partitions import TimeWindowPartitionsDefinition
 from dagster._core.errors import DagsterInvalidDefinitionError
+from dagster._serdes import whitelist_for_serdes
 
 
-class TimeWindowPartitionMapping(PartitionMapping):
+@whitelist_for_serdes
+class TimeWindowPartitionMapping(PartitionMapping, NamedTuple("_IdentityPartitionMapping", [])):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py:13]: "_TimeWindowPartitionMapping"

File: python_modules/dagster/dagster/_core/definitions/partition_mapping.py [hunk 2]
```diff
@@ -77,7 +79,8 @@ def get_downstream_partitions_for_partition_range(
 
 
 @experimental
-class AllPartitionMapping(PartitionMapping):
+@whitelist_for_serdes
+class AllPartitionMapping(PartitionMapping, NamedTuple("_IdentityPartitionMapping", [])):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/partition_mapping.py:83]: "_AllPartitionMapping"

File: python_modules/dagster/dagster/_core/definitions/partition_mapping.py [hunk 3]
```diff
@@ -99,7 +102,8 @@ def get_downstream_partitions_for_partition_range(
 
 
 @experimental
-class LastPartitionMapping(PartitionMapping):
+@whitelist_for_serdes
+class LastPartitionMapping(PartitionMapping, NamedTuple("_IdentityPartitionMapping", [])):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/partition_mapping.py:106]: "_LastPartitionMapping"

File: python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_graph.py [hunk 4]
```diff
@@ -27,14 +40,21 @@ def asset2(asset0):
...
 
-    asset_graph = AssetGraph.from_assets([asset0, asset1, asset2, asset3])
-    assert asset_graph.all_asset_keys == {asset0.key, asset1.key, asset2.key, asset3.key}
-    assert not asset_graph.is_partitioned(asset0.key)
-    assert asset_graph.is_partitioned(asset1.key)
-    assert asset_graph.have_same_partitioning(asset1.key, asset2.key)
-    assert not asset_graph.have_same_partitioning(asset1.key, asset3.key)
-    assert asset_graph.get_children(asset0.key) == {asset1.key, asset2.key}
-    assert asset_graph.get_parents(asset3.key) == {asset1.key, asset2.key}
+    assets = [asset0, asset1, asset2, asset3]
```
Comment (OwenKephart) [inline python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_graph.py:43]: this is great, I meant to get around to this -- any reason `internal_asset_graph == external_asset_graph` would be expected to evaluate to False at this point?

File: python_modules/dagster/dagster/_core/host_representation/external_data.py [hunk 5]
```diff
@@ -785,12 +787,14 @@ def __new__(
...
         input_name: Optional[str] = None,
         output_name: Optional[str] = None,
+        partition_mapping: Optional[PartitionMapping] = None,
     ):
         return super(ExternalAssetDependency, cls).__new__(
             cls,
             upstream_asset_key=upstream_asset_key,
             input_name=input_name,
             output_name=output_name,
+            partition_mapping=partition_mapping,
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/host_representation/external_data.py:797]: should we have a check that the provided partition mapping is serializable / one of the partition mappings we've whitelisted, and if not, raise an error?

this is a breaking change for some people, I'd imagine. the alternative would be to silently ignore the partition mapping if it wasn't whitelisted, or put in a placeholder like "unserializable partition mapping"

**Review by OwenKephart — CHANGES_REQUESTED** [review]
see comments, but this is pretty close

File: python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py:13]: Thanks for catching

File: python_modules/dagster/dagster/_core/host_representation/external_data.py (hunk 5 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/host_representation/external_data.py:800]: Great point. I added some handling for this. Now we consider non-builtin partition mappings as deprecated and we warn on them at asset construction time.  If you do use a non-builtin partition mapping, we revert to the default in host processes.

**Review by OwenKephart — APPROVED** [review]
awesome

**Pull request merged**